### DISCUSSION TITLE
MQTT subscribe in script

### DISCUSSION THREAD

**Discussion by gck303** [arendst/Tasmota#11605] [Q&A] [answered]
I want to create a script to output lots of information to a screen. I have managed to get so far with the rules, but ran out of space. 

Using the example here has been veryhelpful, but it does not seem to work. 
[https://github.com/arendst/Tasmota/discussions/10272](url)

My script is below. It does not actually get any data from the MQTT subscription. These commands DO work within the rules, so I am at a loss and why my script does not work. 

`rule2 ON mqtt#connected DO Subscribe PowerP4, emon/emonesp/P4 ENDON
 ON mqtt#connected DO Subscribe Outside, home-assistant/outside/1/state ENDON`

Thanks, George
```javascript
>D
p4=0
outside=0

>B
if mqtts>0
then
=#subs
endif
#subs
+>Subscribe p4, emon/emonesp/P4
+>Subscribe outside, home-assistant/outside/1/state

>S
if mqttc>0
then
=#subs
endif
print S: %p4% %outside%
dt [x0y214h240]
dt [x0y107h240]

>G
print G: %p4% %outside%

>E
print E: %p4% %outside%
```

**Comment by gemu2015** [comment, marked as answer]
did you compile with #define SUPPORT_MQTT_EVENT ?

to view the decoding process you may enable  #define DEBUG_MQTT_EVENT
and watch the console messages.

you must check with upd like
\>F
if upd[p4]>0
then
print %p4%
endif

**Reply by gck303** [reply]
Yes, it was compiled with SUPPORT_MQTT_EVENT. I have not used DEBUG_MQTT_EVENT, but will look at this. 

Thanks.

**Reply by gck303** [reply]
I have enabled the DEBUG_MQTT_EVENT and increased the error logging to 4 (more debugging) I must say, I cannot see any more details in the Web console. 

When I use the subscribe function, the following is returned. So, it looks like the tasmota is subscribed. This device is called esp12_4. 

```javascript
21:26:53.078 CMD: subscribe
21:26:53.081 SRC: WebConsole from 192.168.0.38
21:26:53.085 CMD: Group 0, Index 1, Command "SUBSCRIBE", Data ""
21:26:53.094 MQT: stat/esp12_4/RESULT = {"Subscribe":"POWER4,emon/emonesp/P4; OUTSIDE,home-assistant/outside/1/state; "}
```

I have a second ESP12 (called esp12_3) which I have enabled rules, rather than scripts. This ESP12 is subscribed to the same two MQTT messages and successfully displays them to a screen. The rule used is:

```javascript
Rule3 ON mqtt#connected DO Subscribe PowerP4, emon/emonesp/P4 ENDON
   ON Event#PowerP4 DO DisplayText [f1x0y240p17]Ecodan IN: %value%W ENDON
   ON mqtt#connected Subscribe Outside, home-assistant/outside/1/state ENDON
   ON Event#Outside DO DisplayText [f2x0y24p20]Outside: %value%C ENDON
```

When I check the subscribe I get the following. Which confirms the same details as 


```javascript
21:22:56.366 SRC: WebConsole from 192.168.0.38
21:22:56.369 CMD: Group 0, Index 1, Command "SUBSCRIBE", Data ""
21:22:56.376 MQT: stat/esp12_3/RESULT = {"Subscribe":"POWERP4,emon/emonesp/P4; ELEC_CURRENT_RATE,home-assistant/elec_current_rate/1/state; "}
```

PS sorry about all the edits. I wanted to get the code/text displaying better.

**Reply by gemu2015** [reply]
ok something must be wrong with your defines if you do not get a logging message from incoming MQTT.
this message should always display your MQTT in debug mode. (%s your topic and event)
"Script: MQTT Topic %s, Event %s"
as long as this message is missing it can not work.

**Reply by gck303** [reply]
Bingo. I have recompiled and now I am getting a MQTT Script message appearing in the console. The 'event' value matches with the value that is in the MQTT payload. 

I can't get the script to process this message or get its value, but that message is arriving. 

```javascript
07:15:14.031 Script: MQTT Topic emon/emonesp/P4, Event 1782
07:15:14.033 Script: Match MQTT message Topic emon/emonesp/P4 with subscription topic emon/emonesp/P4 and key 
07:15:14.037 Script: Match MQTT message Topic emon/emonesp/P4 with subscription topic home-assistant/outside/1/state and key 
07:15:14.399 S: 0.00 0.00
07:15:15.389 S: 0.00 0.00
07:15:16.383 Script: MQTT Topic home-assistant/outside/1/state, Event -1
07:15:16.387 Script: Match MQTT message Topic home-assistant/outside/1/state with subscription topic emon/emonesp/P4 and key 
07:15:16.392 Script: Match MQTT message Topic home-assistant/outside/1/state with subscription topic home-assistant/outside/1/state and key 
```

**Reply by gemu2015** [reply]
you did not specify a key.
what version do you use ?
you should use the last dev version. in a previous version the keyless subscribe did not work.

**Reply by gck303** [reply]
Thank you. Yes, I am using MQTT messages without a key. 

I was using 9.3.1, the release version. I have just moved to the dev version (9.3.1.2) and the script above works!!!!!!

Thank you again. I am sorry, I really should have but the version in my original question.

**Reply by gck303** [reply]
Is there a limit to the number or length of MQTT subscriptions I can put in a script?

I am experiencing problems trying to save the script when I try and configure more than 8. I have searched the documentation and cannot find anything. 

I am having no end of problem building and saving a script. At some point it will not save and the device resets. I eventually get it to save and then it works. It feels like a formatting or error checking bug. 

What checks/validations take place before it is saved?

I do not want to raise an issue, if it is me. BTW it was the dev branch that was compiled yesterday. 

I have got this code into the device. When I enable the script my tasmota restarts. 

```javascript
>D
tides=0
p3=0
p4=0
outside=0
tank=0
flow=0
return=0
ha=0
mn=""
sunrise=0
sunset=0
elcu=0
elne=0
temp=0
humd=0
press=0
dew=0

>B
->setoption64 1
->displaydimmer 15
if mqtts>0
then
=#subs
endif

#subs
+>Subscribe p3, emon/emonesp/P3
+>Subscribe p4, emon/emonesp/P4
;+>Subscribe flow, tele/ecodan/flow_temp
;+>Subscribe return, tele/ecodan/return_temp
+>Subscribe outside, ha/outside
+>Subscribe tank, ha/tank
+>Subscribe sunrise, ha/sunrise
+>Subscribe sunset, ha/sunset
+>Subscribe mn, ha/moon
+>Subscribe elcu, ha/eleccurr
+>Subscribe elne, ha/elecnext

>T
temp=BME680#Temperature
press=BME680#Pressure
humd=BME680#Humidity
dew=BME680#DewPoint

>S
if mqttc>0
then
=#subs
endif
print S: pow %p3% out %outside%

>F
if upd[p3]>0
then
dt [x0y214h240]
dt [x0y107h240]
endif 
```

**Reply by gemu2015** [reply]
what is the restart reason ?
i assume due to the number of Subscribe cmds tasmota is blocked too long.
try to subscribe per console more than 8 and if this works you must backlog the cmds.
here is one example code how to do that.

\>D
cnt=-1
tmp=0

\>B
; put all lines into a text array, keep in mind however that the complete index string size is currently limited to 255 chars
tmp=is(0 "p3,emon/emonesp/P3|p4,emon/emonesp/P4|flow,tele/ecodan/flow_temp|return,tele/ecodan/return_temp|outside,ha/outside|tank,ha/tank|sunrise,ha/sunset|mn,ha/moon|elcu,ha/eleccurr|elne,ha/elecnext|")

if mqtts>0 {
cnt=1
}

\>S

if mqttc>0 {
cnt=1
}

;issue each cmd every second
if cnt>=1 {
+> subscribe %is[cnt]%
cnt+=1
if cnt>=11 {
cnt=-1
}
}

you have several options to send values to a device.
1. subscribe as you did.
2. send tasmota cmds to set the variable directly  via e.g.  script>p4=123
3. use global variables to interchange variables between devices.
